plugin/src/main/groovy/org/openbakery/carthage/CarthageUpdateTask.groovy

```diff
@@ -1,71 +1,109 @@
+	static final String ARG_PLATFORM = "--platform"
+	static final String CARTHAGE_FILE = "Cartfile"
+	static final String CARTHAGE_FILE_RESOLVED = "Cartfile.resolved"
+	static final String CARTHAGE_PLATFORM_IOS = "iOS"
+	static final String CARTHAGE_PLATFORM_MACOS = "Mac"
+	static final String CARTHAGE_PLATFORM_TVOS = "tvOS"
+	static final String CARTHAGE_PLATFORM_WATCHOS = "watchOS"
+	static final String CARTHAGE_USR_BIN_PATH = "/usr/local/bin/carthage"
 
-	public CarthageUpdateTask() {
+	CarthageUpdateTask() {
 		super()
+		newOutputDirectory()
 		setDescription "Installs the carthage dependencies for the given project"
 	}
 
-	@TaskAction
-	void update() {
-
-		def carthagePlatform = getCarthagePlatform()
-		logger.info('Update Carthage for platform ' + carthagePlatform)
-
-		File carthageDirectory = new File(project.projectDir, "Carthage")
-		File platformDirectory = carthagePlatform != 'all' ? new File(new File(carthageDirectory, 'Build'), carthagePlatform) : null
-		if (carthageDirectory.exists() && (platformDirectory != null && platformDirectory.exists())) {
-			logger.info('Skip Carthage update')
-			return
+	@InputFile
+	@Optional
+	@PathSensitive(PathSensitivity.RELATIVE)
+	Provider<File> getCartFile() {
+		// Cf https://github.com/gradle/gradle/issues/2016
+		File file = project.rootProject.file(CARTHAGE_FILE)
+		return project.provider {
+			file.exists() ? file
+					: File.createTempFile(CARTHAGE_FILE, "")
 		}
+	}
 
-		def carthageCommand = getCarthageCommand()
-
-		def output = services.get(StyledTextOutputFactory).create(CarthageUpdateTask)
-		commandRunner.run(
-				project.projectDir.absolutePath,
-				[carthageCommand, "update", "--platform", carthagePlatform],
-				new ConsoleOutputAppender(output))
+	@InputFile
+	@Optional
+	@PathSensitive(PathSensitivity.RELATIVE)
+	Provider<File> getCartResolvedFile() {
+		// Cf https://github.com/gradle/gradle/issues/2016
+		File file = project.rootProject.file(CARTHAGE_FILE_RESOLVED)
+		return project.provider {
+			file.exists() ? file
+					: File.createTempFile(CARTHAGE_FILE_RESOLVED, "resolved")
+		}
 	}
 
-	String getCarthagePlatform() {
+	@Input
+	String getCarthagePlatformName() {
 		switch (project.xcodebuild.type) {
-			case Type.iOS: return 'iOS'
-			case Type.tvOS: return 'tvOS'
-			case Type.macOS: return 'Mac'
-			case Type.watchOS: return 'watchOS'
+			case Type.iOS: return CARTHAGE_PLATFORM_IOS
+			case Type.tvOS: return CARTHAGE_PLATFORM_TVOS
+			case Type.macOS: return CARTHAGE_PLATFORM_MACOS
+			case Type.watchOS: return CARTHAGE_PLATFORM_WATCHOS
 			default: return 'all'
 		}
 	}
 
+	@OutputDirectory
+	Provider<File> getOutputDirectory() {
+		return project.provider {
+			project.rootProject.file("Carthage/Build/" + getCarthagePlatformName())
+		}
+	}
+
+	@TaskAction
+	void update() {
+		if (hasCartFile()) {
+			logger.info('Update Carthage for platform ' + carthagePlatformName)
+
+			def output = services.get(StyledTextOutputFactory).create(CarthageUpdateTask)
```

Comment:
So no skip if the `/Carthage` folder exists?

Reply:
Gradle will automatically skip the task if the current status of the inputs (`cartfile` and `cartfile.resolved`) and the output folder are still in sync with the task cache.
cf https://docs.gradle.org/current/userguide/more_about_tasks.html for more details